sw/device/silicon_creator/manuf/base/ft_personalize.c

```diff
@@ -438,69 +660,89 @@ static status_t personalize_endorse_certificates(ujson_t *uj) {
    * 3. CDI_1 cert
    * 4. Provision Extension certs
    ****************************************************************************/
-  size_t ec_i = 0;
-  size_t ac_i = 0;
-  size_t size = 0;
-
+  // We start scanning the received buffer from zero, the assumption is that the
+  // endorsed UDS cert is the first certificate in the buffer (even if preceeded
+  // by other types of objects).
+  perso_blob_from_host.next_free = 0;
+  // Location where the next cert will be copied.
+  uint8_t *next_cert = all_certs;
+  // How much room left in the destination buffer.
+  size_t free_room = sizeof(all_certs);
+  // Helper structure caching certificate information.
+  cert_block_t block;
   // UDS cert.
-  size = cert_x509_asn1_decode_size_header(endorsed_certs.certs);
-  memcpy(all_certs, endorsed_certs.certs, size);
-  ec_i += size;
-  ac_i += size;
-
-  // CDI_0 cert.
-  size = cert_x509_asn1_decode_size_header(&tbs_certs.certs[cdi_0_offset]);
-  memcpy(&all_certs[ac_i], &tbs_certs.certs[cdi_0_offset], size);
-  ac_i += size;
-
-  // CDI_1 cert.
-  size = cert_x509_asn1_decode_size_header(&tbs_certs.certs[cdi_1_offset]);
-  memcpy(&all_certs[ac_i], &tbs_certs.certs[cdi_1_offset], size);
-  ac_i += size;
-
-  // Remaining (endorsed) certs.
-  size = cert_x509_asn1_decode_size_header(&endorsed_certs.certs[ec_i]);
-  while (size > 0) {
-    memcpy(&all_certs[ac_i], &endorsed_certs.certs[ec_i], size);
-    ec_i += size;
-    ac_i += size;
-    OT_ASSERT_MEMBER_SIZE(manuf_certs_t, certs, 4096);
-    size = ec_i < 4096
-               ? cert_x509_asn1_decode_size_header(&endorsed_certs.certs[ec_i])
-               : 0;
+  TRY(extract_next_cert(&next_cert, &free_room));
+
+  // Now the two CDI certs which were endorsed locally and sent to the host
+  // earlier.
+  size_t cdis[] = {cdi_0_offset, cdi_1_offset};
+  for (size_t i = 0; i < ARRAYSIZE(cdis); i++) {
+    size_t offset = cdis[i];
+    TRY(set_cert_block(perso_blob_to_host.body + offset,
+                       sizeof(perso_blob_to_host.body) - offset, &block));
+    if (block.wrapped_cert_size > free_room)
+      return RESOURCE_EXHAUSTED();
+
+    memcpy(next_cert, block.wrapped_cert_p, block.wrapped_cert_size);
+    LOG_INFO("Copied %s certificate", block.name);
+    next_cert += block.wrapped_cert_size;
+    free_room -= block.wrapped_cert_size;
   }
 
+  // Now the rest of endorsed certificates received from the host, if any.
+  while (perso_blob_from_host.num_objs)
+    TRY(extract_next_cert(&next_cert, &free_room));
+
   /*****************************************************************************
    * Save Certificates to Flash.
    ****************************************************************************/
-  size_t all_certs_offset = 0;
+  // This is where the certificates to be copied are stored, each one wrapped in
+  // with its name prepended to it.
+  uint8_t *certs = all_certs;
```

Comment:
Why reassign the all_certs buffer pointer to another pointer? can you not use `all_certs` pointer directly?